zebra-chain/src/block/merkle.rs

```diff
@@ -1,51 +1,132 @@
+///
+/// This vulnerability does not apply to Zebra, because it does not store invalid
+/// data on disk, and because it does not permanently fail blocks or use an
+/// aggressive anti-DoS mechanism.
+#[derive(Clone, Copy, Eq, PartialEq, Serialize, Deserialize)]
+#[cfg_attr(any(test, feature = "proptest-impl"), derive(Arbitrary))]
+pub struct Root(pub [u8; 32]);
 
-impl<Transaction> ZcashSerialize for Tree<Transaction> {
-    fn zcash_serialize<W: io::Write>(&self, _writer: W) -> Result<(), io::Error> {
-        unimplemented!();
+impl fmt::Debug for Root {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_tuple("Root").field(&hex::encode(&self.0)).finish()
     }
 }
 
-impl<Transaction> ZcashDeserialize for Tree<Transaction> {
-    fn zcash_deserialize<R: io::Read>(_reader: R) -> Result<Self, SerializationError> {
-        unimplemented!();
+fn hash(h1: &[u8; 32], h2: &[u8; 32]) -> [u8; 32] {
+    let mut w = sha256d::Writer::default();
+    w.write_all(h1).unwrap();
+    w.write_all(h2).unwrap();
+    w.finish()
+}
+
+impl<T> std::iter::FromIterator<T> for Root
+where
+    T: std::convert::AsRef<Transaction>,
+{
+    fn from_iter<I>(transactions: I) -> Self
+    where
+        I: IntoIterator<Item = T>,
+    {
+        transactions
+            .into_iter()
+            .map(|tx| tx.as_ref().hash())
+            .collect()
     }
 }
 
-/// A SHA-256d hash of the root node of a merkle tree of SHA256-d
-/// hashed transactions in a block.
-#[derive(Clone, Copy, Eq, PartialEq, Serialize, Deserialize)]
-#[cfg_attr(any(test, feature = "proptest-impl"), derive(Arbitrary))]
-pub struct Root(pub [u8; 32]);
+impl std::iter::FromIterator<transaction::Hash> for Root {
+    fn from_iter<I>(hashes: I) -> Self
+    where
+        I: IntoIterator<Item = transaction::Hash>,
+    {
+        let mut hashes = hashes.into_iter().map(|hash| hash.0).collect::<Vec<_>>();
 
-impl From<Tree<Transaction>> for Root {
-    fn from(merkle_tree: Tree<Transaction>) -> Self {
-        let mut hash_writer = sha256d::Writer::default();
-        merkle_tree
-            .zcash_serialize(&mut hash_writer)
-            .expect("Sha256dWriter is infallible");
-        Self(hash_writer.finish())
+        while hashes.len() > 1 {
+            hashes = hashes
+                .chunks(2)
+                .map(|chunk| match chunk {
+                    [h1, h2] => hash(h1, h2),
+                    [h1] => hash(h1, h1),
+                    _ => unreachable!("chunks(2)"),
+                })
+                .collect();
+        }
+
+        Self(hashes[0])
     }
 }
 
-impl fmt::Debug for Root {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        f.debug_tuple("Root").field(&hex::encode(&self.0)).finish()
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use crate::{block::Block, serialization::ZcashDeserialize};
+
+    #[test]
+    fn block_test_vectors() {
+        for (_, block_bytes) in zebra_test::vectors::MAINNET_BLOCKS.iter() {
```

Comment:
We can check all the mainnet and testnet block test vectors here.

(Since we're not using the network as an input to the check, we can use the combined mainnet and testnet block list.)
```suggestion
        for (_, block_bytes) in zebra_test::vectors::BLOCKS.iter() {
```

Reply:
Amended in a fixup commit.